File: server/src/main/java/io/microconfig/server/vault/VaultConfig.java
package io.microconfig.server.vault;

import io.microconfig.server.rest.exceptions.BadRequestException;
import io.microconfig.server.vault.credentials.KubernetesTokenCredentials;
import io.microconfig.server.vault.credentials.VaultAppRoleCredentials;
import io.microconfig.server.vault.credentials.VaultCredentials;
import io.microconfig.server.vault.credentials.VaultTokenCredentials;
import lombok.Data;

import java.util.Map;

@Data
public class VaultConfig {
    private final String address;
    private final VaultCredentials credentials;

    public static VaultConfig vaultConfig(Map<String, String> config) {
        var address = getValue(config, "microconfig.vault.address");

        var credentials = switch (getValue(config, "microconfig.vault.auth")) {
            case "kubernetes" -> kubernetes(address, config);
            case "token" -> token(address, config);
            case "approle" -> approle(address, config);
            default -> throw new IllegalStateException("Unsupported auth type");
        };

        return new VaultConfig(address, credentials);
    }

    private static VaultCredentials kubernetes(String address, Map<String, String> config) {
        var path = getValue(config, "microconfig.vault.kubernetes.path");
        var role = getValue(config, "microconfig.vault.kubernetes.role");
        var jwt = getValue(config, "microconfig.vault.kubernetes.jwt");
        return new KubernetesTokenCredentials(address, path, role, jwt);
    }

    private static VaultCredentials token(String address, Map<String, String> config) {
        var token = getValue(config, "microconfig.vault.token");
        return new VaultTokenCredentials(token);
    }

    private static VaultCredentials approle(String address, Map<String, String> config) {
        var path = getValue(config, "microconfig.vault.approle.path");
        var role = getValue(config, "microconfig.vault.approle.role");
        var secret = getValue(config, "microconfig.vault.approle.secret");
        return new VaultAppRoleCredentials(address, path, role, secret);
    }

    private static String getValue(Map<String, String> config, String key) {
        var value = config.get(key);
        if (value == null) throw new BadRequestException("Vault config missing: " + key);
        return value;
    }
}

File: server/src/main/kotlin/io/microconfig/server/vault/VaultConfig.kt
package io.microconfig.server.vault

import io.microconfig.server.vault.credentials.KubernetesTokenCredentials
import io.microconfig.server.vault.credentials.VaultAppRoleCredentials
import io.microconfig.server.vault.credentials.VaultCredentials
import io.microconfig.server.vault.credentials.VaultTokenCredentials
import io.microconfig.server.vault.exceptions.VaultAuthException
import java.time.Duration
import java.time.Duration.ofSeconds

data class VaultConfig(val address: String, val timeout: Duration, val credentials: VaultCredentials)

const val DEFAULT_TIMEOUT = 2L

fun vaultConfig(config: Map<String, String>): VaultConfig {
    val address = getValue(config, "microconfig.vault.address")
    val timeout = timeout(config)
    val auth = auth(address, timeout, config)
    return VaultConfig(address, timeout, auth)
}

private fun timeout(config: Map<String, String>): Duration {
    val t = config["microconfig.vault.timeout"]?.toLong() ?: DEFAULT_TIMEOUT
    return ofSeconds(t)
}

private fun auth(address: String, timeout: Duration, config: Map<String, String>): VaultCredentials {
    return when (val auth = getValue(config, "microconfig.vault.auth")) {
        "token" -> token(config)
        "kubernetes" -> kubernetes(address, timeout, config)
        "approle" -> approle(address, timeout, config)
        else -> throw VaultAuthException("Unsupported auth type: $auth")
    }
}

private fun token(config: Map<String, String>): VaultCredentials {
    val token = getValue(config, "microconfig.vault.token")
    return VaultTokenCredentials(token)
}

private fun kubernetes(address: String, timeout: Duration, config: Map<String, String>): VaultCredentials {
    val path = getValue(config, "microconfig.vault.kubernetes.path")
    val role = getValue(config, "microconfig.vault.kubernetes.role")
    val jwt = getValue(config, "microconfig.vault.jwt")
    return KubernetesTokenCredentials(address, timeout, path, role, jwt)
}

private fun approle(address: String, timeout: Duration, config: Map<String, String>): VaultCredentials {
    val path = getValue(config, "microconfig.vault.approle.path")
    val role = getValue(config, "microconfig.vault.approle.role")
    val secret = getValue(config, "microconfig.vault.approle.secret")
    return VaultAppRoleCredentials(address, timeout, path, role, secret)
}

private fun getValue(config: Map<String, String>, key: String): String {
    return config[key] ?: throw VaultAuthException("Vault config missing: $key")
}